### PR TITLE
Running on hardware :). Flash optimised, using better FSMC settings. …

### PR DESCRIPTION
**Pull request by ginge**
…power management glitches all sorted (apart from UART8). Display, flash, pm and logging all have mutexes (ISR detecting for log and pm). Logging optimised to be atomic so we don't twiddle the clock on and off continually. backlight and fading fixed. various small tweaks. Display super wicked fast (but can be better)
removed driver.x. It was annoying

### CODE REVIEW

**Comment by ishotjr** [comment]
:heart_eyes: :rocket:

File: hw/platform/snowy/snowy_rtc.c [hunk 1]
```diff
@@ -260,7 +260,7 @@ void RTC_WKUP_IRQHandler(void)
     if(RTC_GetITStatus(RTC_IT_WUT) != RESET)
     {
         RTC_ClearITPendingBit(RTC_IT_WUT);
-        DRV_LOG("RTC", APP_LOG_LEVEL_DEBUG, "RTC WAKE IRQ");
+        printf("RTC WAKE IRQ");
```
Comment (jwise) [inline hw/platform/snowy/snowy_rtc.c]: any reason why DRV_LOG doesn't work here?

File: rcore/appmanager.c [hunk 2]
```diff
@@ -616,7 +617,7 @@ static void _appmanager_app_thread(void *parms)
             appHeapInit(MAX_APP_MEMORY_SIZE - (MAX_APP_STACK_SIZE * 4), app_stack_heap.byte_buf);
              
             uint32_t *stack_entry = &app_stack_heap.word_buf[(MAX_APP_MEMORY_SIZE / 4) - MAX_APP_STACK_SIZE];
-             
+             printf("LAUNCHING\n");
```
Comment (jwise) [inline rcore/appmanager.c]: 🚀

File: hw/platform/snowy/snowy_rtc.c (hunk 1 above)
Comment (ginge) [reply hw/platform/snowy/snowy_rtc.c]: nope, thats an artifact of trying to fixup logging. Will revert

File: rcore/backlight.c [hunk 3]
```diff
@@ -57,13 +57,15 @@ void rcore_backlight_init(void)
 // use the backlight as additional alert by flashing it
 void rcore_backlight_on(uint16_t brightness_pct, uint16_t time)
 {
-    backlight_message_t msg;
+    // create the message, malloc some memory for it. 
+    // We then add the pointer to this message into the queue
+    backlight_message_t *msg = malloc(sizeof(backlight_message_t));
```
Comment (jwise) [inline rcore/backlight.c]: this is more correct than stacking it but also we should endeavor to not dynamically allocate at all here if we can avoid it

File: rcore/appmanager.c (hunk 2 above)
Comment (ginge) [reply rcore/appmanager.c]: Noted

File: rcore/backlight.c (hunk 3 above)
Comment (ginge) [reply rcore/backlight.c]: ideally at some point i'll change it to use a dma lookup table for the fadeout. Most of this mess can go then as we can switch to a single semaphore and a single message. No need to post the on time and whatnot through a queue. Simplify

File: rcore/log.c [hunk 4]
```diff
@@ -34,58 +35,91 @@ void log_printf_to_ar(const char *layer, const char *module, uint8_t level, cons
...
     // XXX: this means that log_printf *must* be called first from a non-threaded context, for this check is not thread-safe!
     // XXX: verify this here.
     if (_log_mutex == NULL)
         _log_mutex = xSemaphoreCreateMutexStatic(&_log_mutex_buf);
     
-    xSemaphoreTake(_log_mutex, portMAX_DELAY);
+    if(is_interrupt_set())
+    {
+        xHigherPriorityTaskWoken = pdFALSE;
+        xSemaphoreTakeFromISR(_log_mutex, &xHigherPriorityTaskWoken);
```
Comment (jwise) [inline rcore/log.c]: hmm, is this safe?  are ISRs allowed to block?  (how does that even work?)

**Comment by jwise** [comment]
Looks good to me.  :100:

File: rcore/log.c (hunk 4 above)
Comment (ginge) [reply rcore/log.c]: Should be safe. We interrogate the ICSR VECTACTIVE mask to see if we in an ISR. If we detect an ISR call, we use the _FromISR versions from FreeRTOS. They are non-blocking, so in fact the call fails. Should likely detect this failure and do something about it, but it isn't critical, and in fact there are few calls to logging from ISRs

Comment (ginge) [reply rcore/log.c]: Should be safe, yeah. We interrogate the ICSR VECTACTIVE mask to see if there the call is from an ISR. Then we call the _FromISR  version of the function. This is non-blocking (that would be bad mmkay). The bug I see here if that the failure mode of getting the semaphore isn't handled. In this case, it isn't a biggy. we silently fail and get some text thLogEntry!at may be b0rked. XXX fixme TODO I suppose

**Comment by LordChronos123** [comment]
So does this mean the battery finally lasts longer than the previous 6 hours?

**Pull request merged**